automation/JSInstrumentation.py

```diff
@@ -0,0 +1,204 @@
+                    if v != new_logSettings[k]:
+                        print(v)
+                        print(new_logSettings[k])
+                        raise RuntimeError(
+                            f'Mismatching logSettings for object {obj}')
+
+    merged_list = list(merged_map.values())
+    # Make sure list logSettings are unique
+    for request in merged_list:
+        for logSetting in list_log_settings:
+            list_setting_value = request['logSettings'][logSetting]
+            if list_setting_value is None:
+                continue
+            else:
+                if None in list_setting_value:
+                    raise RuntimeError(
+                        f'Mismatching logSettings for object {obj}')
+                else:
+                    # Dedupe
+                    request['logSettings'][logSetting] = list(
+                        set(request['logSettings'][logSetting])
+                    )
+    return merged_list
+
+
+def _handle_obj_string(obj_string):
+    if obj_string.startswith('window'):
+        obj = obj_string
+        instrumentedName = obj_string
+    else:
+        obj = f"window['{obj_string}'].prototype"
+        instrumentedName = obj_string
+    return obj, instrumentedName
+
+
+def build_object_from_request(request):
+    """
+    We need to build a valid object from each request.
+
+    The item may be a string or an object with one key.
+    If the item is a string:
+      - Is it a shortcut e.g. "fingerprinting"
+      - Is it an object (verified by mdn-browser-compat)
+      - If neither, reject
+    If the item is an object:
+      - Must only have one property
+      - The value may be a list or a LogSettings object
+      - If the value is a list, it is transformed into the
+        propertiesToInstrument property of a new LogSettings object.
+    We must also create the instrumentedName value.
+    """
+    obj = None
+    instrumentedName = None
+    logSettings = get_default_log_settings()
+    if isinstance(request, str):
+        obj, instrumentedName = _handle_obj_string(request)
+    elif isinstance(request, dict):
+        assert len(request.keys()) == 1
+        req = list(request.keys())[0]
+        props = request[req]
+        obj, instrumentedName = _handle_obj_string(req)
+        if isinstance(props, list):
+            logSettings['propertiesToInstrument'] = props
+        elif isinstance(props, dict):
+            for k, v in props.items():
+                logSettings[k] = v
+        else:
+            raise RuntimeError('Invalid settings for object')
+    else:
+        raise RuntimeError('Invalid input')
+    return {
+        'object': obj,
+        'instrumentedName': instrumentedName,
+        'logSettings': logSettings
+    }
+
+
+def convert_browser_params_to_js_string(js_instrument_modules):
+    """
+    We accept a list. From the list we need to parse each item.
+    Examples of the requests we accept.
+    // Shortcut
+    "fingerprinting",
```

Comment:
I can imagine we may want to create a "storage" shortcut in the future that instruments localStorage, sessionStorage, indexedDB, Cache, and the like. Since this accepts both shortcut literals and API literals, this might get a bit confusing. E.g., `storage` and `Storage` would do very different things.

What do you think about making it explicit for the user by adding a prefix to the shortcuts? Perhaps something like `[shortcut]-fingerprinting`, `[shortcut]-storage`, `[shortcut]-canvas`, ...? That way we can check for prefix directly and not have to guess or worry about conflicts.

Reply:
We decided on `collection-*`

Reply:
Resolved in 8ac03dc